File: lab1/src/main/java/Service/ReservationService.java
package Service;

import Domain.Reservation;
import Domain.Trip;
import Errors.ValidationError;
import Gateway.ReservationGateway;
import Validator.ReservationValidator;

import java.util.Vector;

public class ReservationService extends BaseService<Reservation, ReservationGateway> {
  public ReservationService(ReservationGateway gateway, ReservationValidator validator) {
    super(gateway, validator);
  }

  public Reservation insert(Reservation reservation) throws ValidationError {
    int seats = countReservedSeatsOnTrip(reservation.getTrip());
    if(18 - seats < reservation.getSeatsNr()) {
      throw new ValidationError("Not enough seats!");
    }
    return super.insert(reservation);
  }

  public Vector<Reservation> getReservationsByTrip(Trip trip) {
    return super.getGateway().getReservationsByTrip(trip);
  }

  public int countReservedSeatsOnTrip(Trip trip) {
    Vector<Reservation> reservations = super.getGateway().getReservationsByTrip(trip);
    int seats = 0;
    for(Reservation reservation: reservations) {
      seats+=reservation.getSeatsNr();
    }
    return seats;
  }

  public int countRemainingSeatsOnTrip(Trip trip) {
    return 18 - countReservedSeatsOnTrip(trip);
  }
}

File: proiect_c#/Services/Service/ReservationService.cs
using csharp.Domain;
using System.Collections.Generic;

public class ReservationService : BaseService<Reservation, IReservationRepository>
{
	public ReservationService(ReservationRepository repository, ReservationValidator validator) : base(repository, validator)
	{
	}

	public new Reservation Insert(Reservation reservation)
	{
		int seats = CountReservedSeatsOnTrip(reservation.Trip);
		if (18 - seats < reservation.SeatsNr)
		{
			throw new ValidationError("Not enough seats!");
		}
		return base.Insert(reservation);
	}

	public List<Reservation> GetReservationsByTrip(Trip trip)
	{
		return repository.GetReservationsByTrip(trip);
	}

	public int CountReservedSeatsOnTrip(Trip trip)
	{
		List<Reservation> reservations = repository.GetReservationsByTrip(trip);
		int seats = 0;
		foreach (var reservation in reservations)
		{
			seats += reservation.SeatsNr;
		}
		return seats;
	}

	public int CountRemainingSeatsOnTrip(Trip trip)
	{
		return 18 - CountReservedSeatsOnTrip(trip);
	}
}